docs/src/design/changing_the_primal.md

```diff
@@ -0,0 +1,435 @@
+  return y, pb
+end
+```
+This is pretty clean now.
+
+
+Our `augmented_primal` is now with-in spitting distance of `rrule`.
+All that is left is a rename, and some extra conventions around multiple outputs and gradients with respect to callable objects.
+
+This has been a journey into how we get to [`rrule`](@ref) as it is defined in `ChainRulesCore`.
+We started with an unaugmented primal function and a `pullback_at` function that only saw the inputs and outputs of the primal.
+We realized a key limitation of this was that we couldn't share computational work between the primal and and gradient passes.
+To solve this we introduced the notation of a some `intermediate` that is shared from the primal to the pullback.
+We successively improved that idea, first by making it a type that held everything that is needed for the pullback: the `PullbackMemory`.
+Which we then made callable --- so it  was itself the pullback.
+Finally, we replaced that separate callable structure with a closure, which kept everything in one place and made it more convenient.
+
+## More Shared Work Examples
+`sin` and the logistic sigmoid are a nice simple examples of when it is useful to share work between primal and the pullback.
+There are many others though.
+It is actually surprising that in so many cases it is reasonable to write the rules where the only shared information between the primal and the pullback is the primal inputs  (like our original `sin`), or primal outputs (like our original logistic sigmoid).
+Under our formulation above, those primal inputs/outputs are shared information just like any other.
+Beyond this there are a number other decent applications.
+
+### `getindex`
+In Julia (and many other numerical languages) indexing can take many more arguments than simply a couple of integers.
+For example, passing in boolean masking arrays (logical indexing), passing in ranges to do slices, etc.
+Converting them to plain integers, arrays of integers, and ranges is `Base.to_indices` is the first thing that `getindex` does.
+It then re-calls `getindex` with these simpler types to get the result.
+
+The result of pulling back the `getindex` operation is always an array that is all zeros,
+except for the elements that are selected, which are set to the appropriate sensitivities being pulled back.
+So identify which actual positions in the array are being gotten/set is common work to both primal and gradient computations.
+We really don't want to deal with fancy indexing types during the pullback, because there are weird edge cases like indexing in such a way that the same element is output twice (and thus we have 2 sensitivities we need to add to it).
+We can pull the `to_indices` out of the primal computation and remember the plain indexes used, then can reuse them to set gradients during the pullback.
+
+See the [code for this in ChainRules.jl](https://github.com/JuliaDiff/ChainRules.jl/blob/v0.7.49/src/rulesets/Base/indexing.jl)
+
+### `exp(::Matrix)`
```

Comment:
I am working on a page about differentiating matrix functions that this can link to. I think even with that, it makes sense to keep this here.